ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidator.scala

```diff
@@ -110,12 +107,22 @@ class PreExecutingSubmissionValidator[WriteSet](
     Timed.timedAndTrackedFuture(
       metrics.daml.kvutils.submission.validator.fetchInputs,
       metrics.daml.kvutils.submission.validator.fetchInputsRunning,
-      ledgerStateReader
-        .read(inputKeys)
-        .map { values =>
-          assert(inputKeys.size == values.size)
-          inputKeys.zip(values).toMap
+      for {
+        inputValues <- ledgerStateReader.read(inputKeys)
+
+        nestedInputKeys = inputValues.collect {
+          case (Some(value), _) if value.hasContractKeyState =>
+            val contractId = value.getContractKeyState.getContractId
+            DamlStateKey.newBuilder.setContractId(contractId).build
         }
+        nestedInputValues <- ledgerStateReader.read(nestedInputKeys)
```

Comment:
This logic needs testing too.

Reply:
I think this should be guarded by a check for `nestedInputKeys.nonEmpty`. there's no point sending a request for and empty set of nested input keys.

Reply:
@gerolf-da: Looks like this is already handled by the `CachingDamlLedgerStateReaderWithFingerprints` class.